nat-lab/tests/telio.py

```diff
@@ -213,57 +213,76 @@ async def notify_peer_state(
         public_key: str,
         states: List[State],
         paths: List[PathType],
-        new_event_explicitly: bool = False,
     ) -> None:
-        def _get_events() -> Set[PeerInfo]:
-            return {
-                event
-                for event in self._peer_state_events
-                if event.public_key == public_key
-                and event.state in states
-                and event.path in paths
-            }
+        while True:
+            peer = self.get_peer_info(public_key)
+            if peer and peer.path in paths and peer.state in states:
+                return
+            await asyncio.sleep(0.01)
+
+    async def notify_peer_event(
+        self,
+        public_key: str,
+        states: List[State],
+        path: PathType,
+        event: Optional[asyncio.Event] = None,
+    ) -> None:
+        def _get_events() -> List[PeerInfo]:
+            return [
+                peer
+                for peer in self._peer_state_events
+                if peer
+                and peer.public_key == public_key
+                and peer.path == path
+                and peer.state in states
+            ]
 
         old_events = _get_events()
+
+        if event:
+            event.set()
+
         while True:
-            peer = self.get_peer_info(public_key)
-            events = _get_events()
-            if (
-                not new_event_explicitly
-                and (
-                    (peer and peer.path in paths and peer.state in states)
-                    or any(events)
-                )
-            ) or (new_event_explicitly and set(events) - set(old_events)):
-                self._peer_state_events = list(set(self._peer_state_events) - events)
+            new_events = _get_events()[len(old_events) :]
+            if new_events:
                 return
             await asyncio.sleep(0.01)
 
     def get_peer_info(self, public_key: str) -> Optional[PeerInfo]:
         return self._peer_info.get(public_key)
```

Comment:
I think this can be changed to, therefore we wouldn't need `self._peer_info`
```
def get_peer_info(self, public_key: str) -> Optional[PeerInfo]:
    return [
        peer_event
        for peer_event in self._peer_state_events
        if peer_event.public_key == public_key
    ][-1]
```

Reply:
Changed